FGEGraphics/GraphicsHelpers/Models/ShapeGenerators.cs

```diff
@@ -0,0 +1,408 @@
+﻿//
+// This file is part of the Frenetic Game Engine, created by Frenetic LLC.
+// This code is Copyright (C) Frenetic LLC under the terms of a strict license.
+// See README.md or LICENSE.txt in the FreneticGameEngine source root for the contents of the license.
+// If neither of these are available, assume that neither you nor anyone other than the copyright holder
+// hold any right or permission to use this software until such time as the official license is identified.
+//
+
+using FGECore.MathHelpers;
+using FGECore.ModelSystems;
+using System;
+using System.Collections.Generic;
+using OpenTK.Mathematics;
+using FGECore.CoreSystems;
+
+namespace FGEGraphics.GraphicsHelpers.Models;
+
+/// <summary>Generates shapes and returns them as a model.</summary>
+public static class ShapeGenerators
+{
+    /// <summary>
+    /// Generates a 3D sphere model.
+    /// Note that the stacks should be lower than the slices to avoid incorrect normals a good example is 10 stacks and 60 slices.
+    /// </summary>
+    /// <param name="radius"></param>
+    /// <param name="stacks"></param>
+    /// <param name="slices"></param>
+    /// <param name="modelEngine"></param>
+    /// <param name="reverseOrder"></param>
+    /// <returns>The generated model.</returns>
+    public static Model GenerateSphere(float radius, uint stacks, uint slices, ModelEngine modelEngine, bool reverseOrder = false)
+    {
+        // Calculate the number of vertices and indices
+        uint vertexCount = (stacks + 1) * (slices + 1);
+        uint numIndices = stacks * slices * 6;  // 2 triangles per stack/slice, 3 indices per triangle
+
+        List<Vector3> vertices = new((int)vertexCount);
+        List<Vector3> normals = new((int)vertexCount);
+        List<Vector2> texCoords = new((int)vertexCount);
+
+        uint[] indices = new uint[numIndices];
+        int index = 0;
+
+        for (int i = 0; i <= stacks; i++)
```

Comment:
Why the `stacks + 1` and `<= stacks` (and same with slices)? Should either not be +1ing constantly or should be doing it at the start as just `stacks += 1;` if there's a reason it needs to be 1 higher than the user specified

Reply:
Oh now I remember its to fully encompass the sphere actually lemme add it back I remember seeing a line in the middle of the sphere.

Reply:
Fixed on https://github.com/FreneticLLC/FreneticGameEngine/pull/22/commits/e56e8f860b242b6694c17c496156fbdbb0266f37

Reply:
There is an active question here (why are you +1ing everywhere) that does not have a clear answer

Reply:
Here's what happens if you don't do that like I said its to encompass the sphere fully and ensure all vertices exist.

![bandicam 2024-03-08 15-27-28-093](https://github.com/FreneticLLC/FreneticGameEngine/assets/97306922/7225a354-70b1-4dc6-9540-802441fdcb37)

Reply:
But you're increasing all usages here - wouldn't it make more sense to have a `- 1` in the `numIndices` calc so that it actually generates the values the user specified?

Reply:
Here's 10 slacks and 10 slices it looks accurate to me the number of vertices is changeable by adjusting the amount of slacks and slices so I don't understand why this is a problem.

![bandicam 2024-03-08 15-43-27-674](https://github.com/FreneticLLC/FreneticGameEngine/assets/97306922/c225b979-5744-4eef-a887-ebb5cbf98fff)

Reply:
Well right now you input 10 and you get 11, is that not wrong?

Reply:
;/ I made the sphere generator that lets you specify how round it is with calculated normals for lighting. Let's be honest who's gonna say "I put 20 stacks and 20 slices and got 512 vertices when I should've gotten 509!" even though its a perfectly usable smooth sphere.

Reply:
no you cannot just generate wrong values because you don't expect people to notice, especially not when it takes like two minutes to fix the error